web3c/confidential.js

```diff
@@ -10,89 +12,107 @@ const Signer = require('./signer');
-  let self = this;
+class Confidential {
+  constructor (web3, storage, mraebox) {
+    this.keyManager = new KeyManager(web3, storage, mraebox);
+    this.setupConfidentialRpc(web3);
+    this.setupConfidentialContract(web3, storage, mraebox);
+    this.setupWrappedAccounts(web3);
+  }
+
   /**
-   * web3.confidential.Contract behaves like web3.eth.Contract.
-   * @param {Object} abi
-   * @param {String} address
-   * @param {Object} options
-   * @param {String} options.key The longterm key of the contract.
-   * @param {bool}   options.saveSession false to disable storing keys.
+   * Creates methods on the confidential namespace that make requests to the confidential_*
+   * web3c rpc endpoints. For example, one may do `web3c.confidential.getPublicKey(address)`.
    */
-  this.Contract = function ConfidentialContract(abi, address, options) {
-    let c = new web3.eth.Contract(abi, address, options);
-    // Copy the wrapped contract to `this`.
-    Object.assign(this, c);
-    this.__proto__ = c.__proto__;
-
-    // Object.DefineProperty's are not copied otherwise.
-    this.defaultAccount = c.constructor.defaultAccount;
-    this.defaultBlock = c.constructor.defaultBlock || 'latest';
-
-    let instanceProvider = provider;
-
-    let keymanager = self.keyManager;
-    if (options && options.saveSession === false) {
-      keymanager = new KeyManager(web3, undefined, mraebox);
-      instanceProvider = new ConfidentialProvider(keymanager, web3._requestManager);
-    }
-
-    c.setProvider.call(this, instanceProvider);
-
-    // Deployed contracts are instantiated with clone.
-    // This patch keeps those clones confidential.
-    this.clone = () => {
-      return new ConfidentialContract(this.options.jsonInterface, this.options.address, this.options);
-    };
+  setupConfidentialRpc(web3) {
+    let methods = [
+      // Second parameter - the long-term key - is intercepted by the provider.
+      new web3.extend.Method({
+        name: 'getPublicKey',
+        call: 'confidential_getPublicKey',
+        params: 1,
+        inputFormatter: [web3.extend.formatters.inputAddressFormatter],
+        outputFormatter: getPublicKeyOutputFormatter
+      }),
+      new web3.extend.Method({
+        name: 'call',
+        call: 'confidential_call_enc',
+        params: 2,
+        inputFormatter: [
+          web3.extend.formatters.inputCallFormatter,
+          web3.extend.formatters.inputDefaultBlockNumberFormatter
+        ],
+        outputFormatter: callOutputFormatter
+      }),
+    ];
+
+    methods.forEach((method) => {
+      method.setRequestManager(web3._requestManager);
+      method.attachToObject(this);
+    });
+  }
+
+  /**
+   * Creates the confidential `Contract` constructor on the `confidential` namespace.
+   * The `Contract` behaves in the same way as the eth `Contract` object, except
+   * transparently encrypts/decrypts all outbound/inbound requests according
+   * to the web3c spec.
+   */
```

Comment:
include @params in the doc comment

Reply:
Done. Though I would rather have types, e.g. typescript, than non-descriptive comments like this. We could consider when we update to Es6 with @beta38.